pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,567 @@
+	iptables   utiliptables.Interface
+	hostIP     net.IP
+}
+
+// Proxier implements ProxyProvider
+var _ proxy.ProxyProvider = &Proxier{}
+
+// ErrProxyOnLocalhost is returned by NewProxier if the user requests a proxier on
+// the loopback address. May be checked for by callers of NewProxier to know whether
+// the caller provided invalid input.
+var ErrProxyOnLocalhost = fmt.Errorf("cannot proxy on localhost")
+
+// NewProxier returns a new Proxier given an address on which to listen.
+// Because of the iptables logic, It is assumed that there is only a single
+// Proxier active on a machine.
+// An error will be returned if the proxier cannot be started due to an invalid
+// ListenIP (loopback) or if iptables fails to update or acquire the initial lock.
+// Once a proxier is created, it will keep iptables up to date in the background and
+// will not terminate if a particular iptables call fails.
+func NewProxier(ipt utiliptables.Interface) (*Proxier, error) {
+	hostIP, err := util.ChooseHostInterface()
+	if err != nil {
+		return nil, fmt.Errorf("failed to select a host interface: %v", err)
+	}
+
+	glog.V(2).Infof("Setting proxy IP to %v and initializing iptables", hostIP)
+	return &Proxier{
+		serviceMap: make(map[proxy.ServicePortName]*serviceInfo),
+		iptables:   ipt,
+		hostIP:     hostIP,
+	}, nil
+}
+
+func (proxier *Proxier) getServiceInfo(service proxy.ServicePortName) (*serviceInfo, bool) {
+	proxier.mu.Lock()
+	defer proxier.mu.Unlock()
+	info, ok := proxier.serviceMap[service]
+	return info, ok
+}
+
+func (proxier *Proxier) setServiceInfo(service proxy.ServicePortName, info *serviceInfo) {
+	proxier.mu.Lock()
+	defer proxier.mu.Unlock()
+	proxier.serviceMap[service] = info
+}
+
+func (proxier *Proxier) sameConfig(info *serviceInfo, service *api.Service, port *api.ServicePort) bool {
+	if info.protocol != port.Protocol || info.portal.port != port.Port {
+		return false
+	}
+	if !info.portal.ip.Equal(net.ParseIP(service.Spec.ClusterIP)) {
+		return false
+	}
+	if !ipsEqual(info.deprecatedPublicIPs, service.Spec.DeprecatedPublicIPs) {
+		return false
+	}
+	if !api.LoadBalancerStatusEqual(&info.loadBalancerStatus, &service.Status.LoadBalancer) {
+		return false
+	}
+	if info.sessionAffinityType != service.Spec.SessionAffinity {
+		return false
+	}
+	return true
+}
+
+func ipsEqual(lhs, rhs []string) bool {
+	if len(lhs) != len(rhs) {
+		return false
+	}
+	for i := range lhs {
+		if lhs[i] != rhs[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// TODO: what should this be?
+const syncIntervalIptables = 5 * time.Second
+
+// SyncLoop runs periodic work.  This is expected to run as a goroutine or as the main loop of the app.  It does not return.
+func (proxier *Proxier) SyncLoop() {
+	//TODO: should the sync interval differ from Proxier's ?
+	t := time.NewTicker(syncIntervalIptables)
+	defer t.Stop()
+	for {
+		<-t.C
+		glog.V(6).Infof("Periodic sync")
+		if err := proxier.iptablesInit(); err != nil {
```

Comment:
I'm not sure, currently all of syncProxyRules is done with one
iptables-restore call which is in theory atomic, while iptablesInit is not.

On Mon, Aug 10, 2015 at 7:25 PM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/proxy/iptables/proxier.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210#discussion_r36696806>
> :
>
> > +	}
> > +	return true
> > +}
> > +
> > +// TODO: what should this be?
> > +const syncIntervalIptables = 5 * time.Second
> > +
> > +// SyncLoop runs periodic work.  This is expected to run as a goroutine or as the main loop of the app.  It does not return.
> > +func (proxier *Proxier) SyncLoop() {
> > +	//TODO: should the sync interval differ from Proxier's ?
> > +	t := time.NewTicker(syncIntervalIptables)
> > +	defer t.Stop()
> > +	for {
> > +		<-t.C
> > +		glog.V(6).Infof("Periodic sync")
> > +		if err := proxier.iptablesInit(); err != nil {
>
> since iptablesInit() and syncProxyRules() are joined at the hip now, what
> do you think of just flattening them into a single call?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210/files#r36696806>
> .
>